---
title: Fragment size distribution of merged overlapping reads vs. aligned reads
author: Shaun Jackman
date: 2015-01-28
---

+ Species: E.coli K-12 MG1655
+ Data: https://basespace.illumina.com/sample/3756762/Ecoli 
+ Merged overlapping reads using abyss-mergepairs -q 15
+ Aligned reads to an assembly of the same reads using ABySS 1.5.2

```{r load-libraries, message=FALSE}
library(dplyr)
library(ggplot2)
library(knitr)
library(magrittr)
library(fitdistrplus)
library(tidyr)
library(truncnorm)
```

```{r read-data}
merged.data <- read.table("ecoli_merged.fastq.hist",
	col.names = c("Fragment.size", "Count"))
aligned.data <- read.table("k364/pe600-3.hist",
	col.names = c("Fragment.size", "Count"))
data <- rbind(
	merged.data %>% cbind(data.frame(Group = "Merged")),
	aligned.data %>% cbind(data.frame(Group = "Aligned")))
```

```{r summarize-data}
aligned.stats <- aligned.data %$%
	rep.int(x = Fragment.size, times = Count) %>%
	boxplot.stats(coef = 3)
str(aligned.stats)
```

```{r plot-data, message = FALSE}
ggplot(data = data) +
	aes(x = Fragment.size, y = Count, group = Group, colour = Group) +
	geom_point(size = 1) + geom_smooth() +
	xlim(aligned.stats$stats[1], aligned.stats$stats[5])
```

```{r remove-outliers}
aligned.data.omit.outliers <- aligned.data %>%
	filter(Fragment.size >= aligned.stats$stats[1]
		& Fragment.size <= aligned.stats$stats[5])
```

```{r reshape-data}
rep.aligned.data <- aligned.data.omit.outliers %$%
	data.frame(Fragment.size = rep.int(Fragment.size, Count))
```

```{r fit-normal}
aligned.normal.fit <- rep.aligned.data %$% data.frame(
	mean = mean(Fragment.size),
	sd = sd(Fragment.size))
aligned.normal.fit

ggplot(rep.aligned.data %>% sample_n(100000)) +
	aes(x = Fragment.size) +
	geom_histogram(binwidth = 50, aes(y = ..density..)) +
	geom_density(aes(colour = "Empirical")) +
	stat_function(fun = dnorm, aes(colour = "Normal"), args = aligned.normal.fit)
```


```{r fit-truncated-normal}
merged.data.omit.outliers <- merged.data %>%
	filter(Fragment.size >= aligned.stats$stats[1]
		& Fragment.size <= aligned.stats$stats[5])

rep.merged.data <- merged.data.omit.outliers %$%
	data.frame(Fragment.size = rep.int(Fragment.size, Count))

merged.normal.fit <- rep.merged.data %$% data.frame(
	mean = mean(Fragment.size),
	sd = sd(Fragment.size))
merged.normal.fit

merged.truncnorm.fix.arg <- data.frame(
	a = min(rep.merged.data) %>% as.double,
	b = max(rep.merged.data) %>% as.double)
merged.truncnorm.fix.arg

merged.truncnorm.fit <- fitdist(
	rep.merged.data$Fragment.size %>% as.double,
	"truncnorm",
	start = merged.normal.fit,
    fix.arg = merged.truncnorm.fix.arg)
plot(merged.truncnorm.fit)
```

```{r plot-truncated-normal-fit}
ggplot(rep.merged.data %>% sample_n(100000)) +
	aes(x = Fragment.size) +
	geom_histogram(binwidth = 25, aes(y = ..density..)) +
	geom_density(aes(colour = "Empirical")) +
	stat_function(fun = dnorm, aes(colour = "Normal"),
		args = merged.normal.fit) +
	stat_function(fun = dtruncnorm, aes(colour = "Truncated normal"),
		args = c(merged.truncnorm.fit$estimate, merged.truncnorm.fix.arg)) +
	stat_function(fun = dtruncnorm, aes(colour = "Aligned normal"),
		args = c(aligned.normal.fit, merged.truncnorm.fix.arg))
```

```{r tabulate-data}
kable(data)
```
